pam/integration-tests/gdm-integration_test.go

```diff
@@ -0,0 +1,471 @@
+				"invalid password 'really, it's not a goodpass!', should be 'goodpass'",
+			},
+			wantError:       pam.ErrAuth,
+			wantAcctMgmtErr: ignoreError,
+		},
+		"Error on authenticating unknown user": {
+			pamUser: "user-unknown",
+			eventPollResponses: map[gdm.EventType][]*gdm.EventData{
+				gdm.EventType_startAuthentication: {
+					gdm_test.IsAuthenticatedEvent(&authd.IARequest_AuthenticationData_Challenge{
+						Challenge: "",
+					}),
+				},
+			},
+			wantPamErrorMessages: []string{
+				"user not found",
+			},
+			wantError:       pam.ErrAuth,
+			wantAcctMgmtErr: ignoreError,
+		},
+		"Error on invalid fido ack": {
+			pamUser:     "user-mfa",
+			authModeIDs: []string{passwordAuthID, fido1AuthID},
+			eventPollResponses: map[gdm.EventType][]*gdm.EventData{
+				gdm.EventType_startAuthentication: {
+					gdm_test.IsAuthenticatedEvent(&authd.IARequest_AuthenticationData_Challenge{
+						Challenge: "goodpass",
+					}),
+					gdm_test.IsAuthenticatedEvent(&authd.IARequest_AuthenticationData_Wait{}),
+				},
+			},
+			wantPamErrorMessages: []string{
+				fido1AuthID + " should have wait set to true",
+			},
+			wantError:       pam.ErrAuth,
+			wantAcctMgmtErr: ignoreError,
+		},
+	}
+	for name, tc := range testCases {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Cleanup(pam_test.MaybeDoLeakCheck)
+			t.Parallel()
+
+			// We run a daemon for each test, because here we don't want to
+			// make assumptions whether the state of the broker and each test
+			// should run in parallel and work the same way in any order is ran.
+			ctx, cancel := context.WithCancel(context.Background())
+			socketPath, stopped := testutils.RunDaemon(ctx, t, daemonPath,
+				testutils.WithEnvironment(grouptests.GPasswdMockEnv(t, gpasswdOutput, groupsFile)...),
+			)
+			t.Cleanup(func() {
+				cancel()
+				<-stopped
+			})
+			serviceFile := createServiceFile(t, "module-loader", libPath,
+				append(slices.Clone(args), "socket="+socketPath), pamDebugIgnoreError)
+
+			gh := newGdmTestModuleHandler(t, serviceFile, tc.pamUser)
+			t.Cleanup(func() { require.NoError(t, gh.tx.End()) })
+			gh.eventPollResponses = tc.eventPollResponses
+
+			if tc.supportedLayouts == nil {
+				gh.supportedLayouts = []*authd.UILayout{pam_test.FormUILayout()}
+			}
+
+			gh.protoVersion = gdm.ProtoVersion
+			if tc.protoVersion != 0 {
+				gh.protoVersion = tc.protoVersion
+			}
+
+			gh.selectedBrokerName = tc.brokerName
+			if gh.selectedBrokerName == "" {
+				gh.selectedBrokerName = exampleBrokerName
+			}
+
+			gh.selectedAuthModeIDs = tc.authModeIDs
+			if gh.selectedAuthModeIDs == nil {
+				gh.selectedAuthModeIDs = []string{passwordAuthID}
+			}
+
+			authResult := make(chan error)
+			go func() {
+				authResult <- gh.tx.Authenticate(pam.Flags(0))
+			}()
+
+			var err error
+			select {
+			case <-time.After(10 * time.Second):
+				t.Fatal("Authentication timed out!")
+			case err = <-authResult:
+			}
+
+			require.ErrorIs(t, err, tc.wantError)
```

Comment:
please add sentence to the assertion, like "we expected blablabla to return this result") (same for other assertions, I added more below but there are some more to add too)

Reply:
+1 (I've already told you about this in other reviews as well, @3v1n0)

Reply:
Eh, I know... But when I got errors they look very straight forward to follow to me via line numbers, but that's fine :)

Addressing them though.